plumber/cli_server.go

```diff
@@ -89,6 +99,119 @@ func (p *Plumber) RunServer() error {
+	// We will talk gRPC to Foreman, and Foreman will talk gRPC to us
+	foremanConn, err := yamux.Client(conn, yamux.DefaultConfig())
+	if err != nil {
+		p.log.Errorf("couldn't create yamux server: %s", err)
+		return false
+	}
+
+	dialOpts := make([]grpc.DialOption, 0)
+
+	if p.Config.CLIOptions.Server.RemoteControlDisableTls {
+		dialOpts = append(dialOpts, grpc.WithInsecure())
+	} else {
+		dialOpts = append(dialOpts, grpc.WithTransportCredentials(credentials.NewTLS(
+			&tls.Config{
+				InsecureSkipVerify: true,
+			},
+		)))
+	}
+
+	dialOpts = append(dialOpts, grpc.WithContextDialer(func(context.Context, string) (net.Conn, error) {
+		return foremanConn.Open()
+	}))
+
+	// Establish Plumber -> Foreman gRPC connection over our yamux session
+	foremanGRPCConn, err := grpc.Dial(foremanAddr, dialOpts...)
+	if err != nil {
+		p.log.Errorf("failed to create grpc client: %s", err)
+		conn.Close()
+		return false
+	}
+
+	// Register ourselves with the foreman service
+	client := protos.NewForemanClientClient(foremanGRPCConn)
+	authResp, err := client.Register(context.Background(), &protos.RegisterRequest{
+		ApiToken:     p.Config.CLIOptions.Server.RemoteControlApiToken,
+		ClusterId:    p.Config.CLIOptions.Server.ClusterId,
+		PlumberToken: p.Config.CLIOptions.Server.AuthToken,
+		NodeId:       p.Config.CLIOptions.Server.NodeId,
+	})
+	if err != nil {
+		p.log.Errorf("failed to register with remote control server. remote control not available: %s", err)
+		conn.Close()
+		foremanGRPCConn.Close()
+		return false
+	}
+
+	if !authResp.Success {
+		p.log.Errorf("failed to register with remote control server. remote control not available: %s", authResp.Message)
+		conn.Close()
+		foremanGRPCConn.Close()
+		return false
+	}
+
+	// Start a second plumber gRPC server over the yamux session so that
+	// Foreman can send gRPC calls to this Plumber instance
+	if err := p.startGRPCServer(foremanConn, foremanAddr); err != nil {
+		p.log.Fatalf("unable to run remote control gRPC server: %s", err)
+	}
+
+	// Monitor for connection loss and attempt to reconnect
+	go p.watchForForemanDisconnect(foremanConn)
+
+	return true
+}
+
+// watchForForemanDisconnect watches for a disconnect from the foreman plumber control service
+// If it detects a disconnect, it will attempt to continually retry the connection
+func (p *Plumber) watchForForemanDisconnect(sess *yamux.Session) {
+	ch := sess.CloseChan()
+
+	// Block until connection is closed/lost
+	<-ch
+
+	// Try to reconnect w/backoff
+	var i int
+	for {
+		retry := ForemanReconnectPolicy.Duration(i)
+		p.log.Errorf("No connection to remote control server. Remote control not available, attempting "+
+			"reconnect in %s", retry.String())
+		time.Sleep(retry)
+
+		if ok := p.runRemoteControl(); ok {
+			p.log.Info("Reconnected successfully to remote control server")
+			// Reconnected successfully, exit out of this goroutine
+			// Another one will be started for the new connection in runRemoteControl()
+			return
+		}
```

Comment:
What happens when all reconnects (with backoff) fail?

Reply:
@dselans It will continue to retry with the highest interval, in this case, every 5 minutes. Since this will be running in server mode, I think we just want it to retry indefinitely right?